lib/irb/input-method.rb

```diff
@@ -271,100 +280,137 @@ def auto_indent(&block)
-
-      right_x = cursor_pos_to_render.x + autocomplete_dialog.width
-      if right_x + width > screen_width
-        right_width = screen_width - (right_x + 1)
-        left_x = autocomplete_dialog.column - width
-        left_x = 0 if left_x < 0
-        left_width = width > autocomplete_dialog.column ? autocomplete_dialog.column : width
-        if right_width.positive? and left_width.positive?
-          if right_width >= left_width
+        doc = nil
+        used_for_class = false
+        if not name =~ /#|\./
+          found, klasses, includes, extends = driver.classes_and_includes_and_extends_for(name)
+          if not found.empty?
+            doc = driver.class_document(name, found, klasses, includes, extends)
+            used_for_class = true
+          end
+        end
+        unless used_for_class
+          doc = RDoc::Markup::Document.new
+          begin
+            driver.add_method(doc, name)
+          rescue RDoc::RI::Driver::NotFoundError
+            doc = nil
+          rescue
+            return nil # unknown error
+          end
+        end
+        return nil if doc.nil?
+        width = 40
+
+        right_x = cursor_pos_to_render.x + autocomplete_dialog.width
+        if right_x + width > screen_width
+          right_width = screen_width - (right_x + 1)
+          left_x = autocomplete_dialog.column - width
+          left_x = 0 if left_x < 0
+          left_width = width > autocomplete_dialog.column ? autocomplete_dialog.column : width
+          if right_width.positive? and left_width.positive?
+            if right_width >= left_width
+              width = right_width
+              x = right_x
+            else
+              width = left_width
+              x = left_x
+            end
+          elsif right_width.positive? and left_width <= 0
             width = right_width
             x = right_x
-          else
+          elsif right_width <= 0 and left_width.positive?
             width = left_width
             x = left_x
+          else # Both are negative width.
+            return nil
           end
-        elsif right_width.positive? and left_width <= 0
-          width = right_width
+        else
           x = right_x
-        elsif right_width <= 0 and left_width.positive?
-          width = left_width
-          x = left_x
-        else # Both are negative width.
-          return nil
         end
+        formatter = RDoc::Markup::ToAnsi.new
+        formatter.width = width
+        dialog.trap_key = alt_d
+        mod_key = RUBY_PLATFORM.match?(/darwin/) ? "Option" : "Alt"
+        message = "Press #{mod_key}+d to read the full document"
+        contents = [message] + doc.accept(formatter).split("\n")
+        contents = contents.take(preferred_dialog_height)
+
+        y = cursor_pos_to_render.y
+        Reline::DialogRenderInfo.new(pos: Reline::CursorPos.new(x, y), contents: contents, width: width, bg_color: '49')
+      }
+    end
+
+    def display_document(matched, driver: nil)
```

Comment:
Nice extraction 👍 
I'd like to store the driver in `RelineInputMethod` because it doesn't need to be initialised every time we show the doc, and memoise it with lazy initialisation makes testing complicated.

Reply:
For performance, storing driver is not needed here.
`RDoc::RI::Driver.new` takes 0.18sec for the first time and after that, takes 0.035sec.
In here, driver is only initialized when user press TAB key twice to explicitly show document in noautocomplete mode. I think 0.035sec is acceptable.

The problem is more with document dialog of autocomplete mode. It does not cache driver and it takes each 0.035sec while scrolling completion dialog by TAB key.

SHOW_DOC_DIALOG calling RDoc::RI::Driver.new (in branch master)
https://github.com/ruby/irb/blob/1d2d35dd3309924588b9c6b89f3904690734997f/lib/irb/input-method.rb#L291-L293

Creating `RDoc::RI::Driver.new` at startup will slowdown IRB's startup time, so lazily initialize it and use it from both `show_doc_dialog_proc` and `display_document` might be better. (maybe after this pull request?)

Reply:
Good catch on the performance penalty of initialising it repeatedly in dialog 👍 
That said, I don't think `0.035` sec at the startup time would be too big of an issue. Storing it in the constructor makes the dependency easy to understand, which outweighs the minor delay IMO.

But I agree we should deal with this in another PR, and also let's do a proper profiling on IRB startup?